File: project_euler/Problem041/Problem041.py
import math
from typing import List

class Problem041:
    def solution(self) -> None:
        elements = [7, 6, 5, 4, 3, 2, 1]
        result = self.__permutation(elements, 0, len(elements), 0)
        print(result)

    def __permutation(self, elements: List[int], index: int, length: int, pivot: int) -> int:
        if index == length:
            result = 0
            for element in elements:
                result = result * 10 + element
            if self.__is_prime(result):
                return max(result, pivot)
            return pivot

        result = self.__permutation(elements, index + 1, length, pivot)
        for i in range(index + 1, length):
            elements[index], elements[i] = elements[i], elements[index]
            result = self.__permutation(elements, index + 1, length, result)
            elements[index], elements[i] = elements[i], elements[index]
        return result

    def __is_prime(self, value: int) -> bool:
        if 1 != value % 6 and 5 != value % 6:
            return 2 == value or 3 == value

        border = int(math.sqrt(value))
        for i in range(5, border + 1, 6):
            if not value % i or not value % (i + 2):
                return False
        return 1 != value

File: project_euler/Problem041/Problem041.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem041() {
    elements := []int{7, 6, 5, 4, 3, 2, 1}
    result := permutation(elements, 0, len(elements), int64(0))
    fmt.Println(result)
}

func permutation(elements []int, index int, length int, pivot int64) int64 {
    if index == length {
        result := int64(0)
        for _, element := range elements {
            result = result * 10 + int64(element)
        }
        if isPrime(result) {
            return int64(math.Max(float64(result), float64(pivot)))
        }
        return pivot
    }

    result := permutation(elements, index + 1, length, pivot)
    for i := index + 1; i < length; i++ {
        elements[index], elements[i] = elements[i], elements[index]
        result = permutation(elements, index + 1, length, result)
        elements[index], elements[i] = elements[i], elements[index]
    }
    return result
}

func isPrime(value int64) bool {
    if 1 != value % 6 && 5 != value % 6 {
        return 2 == value || 3 == value
    }

    border := int64(math.Sqrt(float64(value)))
    for i := int64(5); i <= border; i += 6 {
        if 0 == value % i || 0 == value % (i + 2) {
            return false
        }
    }
    return 1 != value
}
